File: src/main/jvm/de/rub/bi/inf/openbimrl/functions/ifc/GetGlobalID.java
package de.rub.bi.inf.openbimrl.functions.ifc;

import de.rub.bi.inf.openbimrl.NodeProxy;
import de.rub.bi.inf.openbimrl.functions.AbstractFunction;
import de.rub.bi.inf.openbimrl.functions.annotations.OpenBIMRLFunction;

import java.util.ArrayList;

/**
 * Retrives the GUID of an IFC element.
 *
 * @author Marcel Stepien
 */
@OpenBIMRLFunction
public class GetGlobalID extends AbstractFunction {

    public GetGlobalID(NodeProxy nodeProxy) {
        super(nodeProxy);
    }

    @Override
    public void execute() {

        final var objects = getInputAsCollection(0);
        if (objects.isEmpty()) return;


        final var resultValues = new ArrayList<String>();

        /*for (Object o : objects) {
            if (o instanceof IIFCObject ifcObj) {

                resultValues.add(
                        ifcObj.getGlobalId().getDecodedValue()
                );

            }
        }
*/
        if (resultValues.size() == 1) {
            setResult(0, resultValues.get(0));
        } else {
            setResult(0, resultValues);
        }

    }

}

File: src/main/jvm/de/rub/bi/inf/openbimrl/functions/ifc/GetGlobalID.kt
package de.rub.bi.inf.openbimrl.functions.ifc

import de.rub.bi.inf.nativelib.IfcPointer
import de.rub.bi.inf.openbimrl.NodeProxy
import de.rub.bi.inf.openbimrl.functions.AbstractFunction
import de.rub.bi.inf.openbimrl.functions.annotations.FunctionInput
import de.rub.bi.inf.openbimrl.functions.annotations.FunctionOutput
import de.rub.bi.inf.openbimrl.functions.annotations.OpenBIMRLFunction

@OpenBIMRLFunction(description = "Retrives the GUID of an IFC element.")
class GetGlobalID(nodeProxy: NodeProxy) : AbstractFunction(nodeProxy) {

    @FunctionInput(0, name = "IfcElement_List", collectionType = IfcPointer::class)
    lateinit var ifcElements: Collection<IfcPointer>

    @FunctionOutput(0, name = "ID_List")
    var ids: Any? = null

    override fun execute() {
        if (ifcElements.isEmpty()) return

        val guids = ifcElements.map { it.guid }
        ids = if (guids.size == 1) guids[0] else guids
    }
}
